### ISSUE TITLE
Group Rolls

### ISSUE THREAD

**Issue by GreenImp** [GreenImp/rpg-dice-roller#88]
We should allow rolls to be grouped, so that we can carry out actions on a group at a time rather than individual rolls.

This was discussed briefly in #86 
It would allow us to do things like `{2d6}>4` to only return a success if the total rolled is greater than 4, rather than this, `2d6>4` which will check each roll.

Or, a better example, with multiple rolls is something like this: `{4d6+2d8, 3d20+3, 5d10+1}>40`

## Without modifiers

If a roll group has no modifiers, then the total value of the group will be the totals of each sub-roll added together.

```
{4d6, 2d10, d4}: {[2, 6, 4, 2], [7, 4], [1]} = 26
```

* [x] Add sub-roll values together

## Modifiers

With modifier's the total will still generally be the sum of the sub-rolls, but they will be altered by the modifiers first.

Applicable modifiers (Other's won't work with roll groups):
* [x] Keep
  Keep the best or last sub-roll result values.
  
  * [x] For a single sub-roll, it will keep the specified number of _dice_ out of all the rolls:
    ```
    // roll 4d10 and 5d6, out of those 9 dice, only keep the highest two. Then total the values
    {4d10*5d6}k2: {[8, 1d, 4d, 1d]*[1d, 5, 1d, 5d, 3d]} = 40
    ```
  * [x] For multiple sub-rolls, it will keep the specified number of _sub-rolls_ based on their total value:
    ```
    // roll the sub-rolls, total their values, and only keep the highest two
    {4d10, 5d6, 2d10}k2: {[8, 1, 4, 1], [1, 5, 1, 5, 3], [6, 3]d} = 29
    ```
* [x] Drop
  
  * [x] Single sub-roll
  * [x] Multiple sub-rolls
 
  Same as Keep, but for dropping.
  ```
  // roll the sub-rolls, total them, and drop the lowest
  {4d6+2d8, 3d20+3, 5d10+1}d1: {([3, 5, 3, 2]+[7, 4])d, [2, 14, 7]+3, [5, 7, 3, 9, 8]+1]} = 58
  ```
  Unsure on how to display modifiers for sub-rolls, but perhaps parenthesis like above could work.
* [x] Target success / failure
  * [x] Single sub-roll
  * [x] Multiple sub-rolls
 
  I need to look into how this should work. The functionality on Roll20 seems a bit strange, so not sure we should mimic this exactly. See details below.
* [x] Sorting
  * [x] For a single sub-roll, it will sort each dice roll as though they all have the sorting modifier applied.
  * [x] For multiple sub-rolls, it will also sort the sub-rolls themselves by result value

### Target success / failure

Looking at how Roll20 handle it seems a bit illogical:

<details>
  <summary>Single sub-rolls:</summary>

  > `{3d20+5}>21` - Roll 3 d20's, _for each roll_ add 5 and then count a success for each result of 21 or more.

  So something like:
  ```
  {3d20+5}>21: {[20+5, 7+5, 19+5]} = 2 successes
  ```

  ![group-roll-target-success-single](https://user-images.githubusercontent.com/1846676/80977585-dc089580-8e1c-11ea-9638-14896b850204.png)

  Whereas I would logically expect it to roll 3 d20s and add 5 to the total, count a success if the total is greater than 21:
  ```
  {3d20+5}>21: {[20, 7, 19]+5} = 1 success
  ```
</details>

<details>
  <summary>Multiple sub-rolls</summary>

  This works as expected:

  > `{4d6+2d8, 3d20+3, 5d10+1}>40` - Roll each of the three sub-roll expression and total them up. Count one success for each sub-roll total of 40 or more.

  It seems to handle the `3d20+3` correctly by rolling 3 d20s and adding 3 to the total, as can be seen here:

  If it added 3 to each roll, this would pass:
  ![group-roll-target-success-multiple](https://user-images.githubusercontent.com/1846676/80977726-06f2e980-8e1d-11ea-8c3a-7c8d69fd4905.png)

  But this is the only roll that passes, because it's the only way to reach a value of 7 if it add 3 to the total:
![group-roll-target-success-multiple-success](https://user-images.githubusercontent.com/1846676/80977859-3b66a580-8e1d-11ea-9e1e-43b8a4d532c5.png)
</details>

So there is an inconsistency with how mathematical equations are handled for sub-rolls, depending on whether there is one sub-roll, or multiple, which seems wrong and confusing to me.

**Comment by vinceRV** [comment]
Hello @GreenImp,

Many thanks for your work on this library, I've been using it quite a lot since the start of the lockdown.

This grouping feature is really interesting. I've tried a couple of things with your feature branch but without so much success. I'd be willing to help if I can.

It's a bit hacky but as a workaround for some simple cases, I've forked your project and added Math.max to the set of supported functions. Please, let me know if it's worth making a pull request.

Thanks again !

**Comment by GreenImp** [comment]
Hey, I'm glad you're finding the library useful 😀

I'm always happy for some help. I'd really like to get this functionality complete, but I've not had the chance to look at it properly.

If you have any code samples of what you've tried, or have any suggestions, let me know!
Also happy to look at any pull requests if you get a working example.

I'm going to look at the group stuff again this week and see if I can figure it a nice way of handling it. Any ideas are more than welcome!

**Comment by vinceRV** [comment]
A way to go might be to implement RollGroup.roll(), as you've started doing and then reverted, which would roll the sub expressions in a similar way as what's done in StandardDice.roll().

What should also be implemented is runing what's in RollGroups._modifiers. This requires adding a _modifiers array in RollResults to store the outputs, same as in RollResult. And modify toString(). Maybe, I would replace the first parameter of *Modifier.run() with the array of rolls instead of the RollResults instance. It seems to me that the latter is not really needed because the modifiers are mostly interested in the rolls property.

Then, implement todos here and there like in DiceRoll.output and _calculate.

I must admit that I haven't spent a lot of time digging. I've started this way and got lost in the tests. Your coverage is really impressive.

**Comment by GreenImp** [comment]
Thanks!

I started looking at this again yesterday. I think I've got a way forward with it and I should make some further progress this week.

**Comment by GreenImp** [comment]
Does anyone following this ticket have any input as to how the "Target success / failure" modifier should work on roll groups?

I've updated the ticket desciption to detail my findings on how it seems to work on Roll20. I'm not convinced that this library should handle it in the same way, as it seems inconsistent and confusing.

I would appreciate some feedback from anyone who has an interested in this functionality 😄

**Comment by vinceRV** [comment]
For me, it's also common sense that the result of a+b is the addition of the two operands a and b.
I would find it very confusing to have a different behaviour given b is a plain number or a dice pool, be it in a group roll or not.

So, personnally, I would also go with:
1. {3d20+5}>21: {[20, 7, 19]+5*} = 1 success (ie 20+7+19+5>21)
2. {3d20+2d100}>21: {[20, 7, 19]+[5, 22]*} = 1 success (ie 20+7+19+5+22>21)
3. {3d20+5,2d100}>21: {[20, 7, 19]+5*, [5, 22*]} = 2 successes (ie 20+7+19+5>21, 5<21 and 22>21)

What I still find a bit confusing with this are the results. In the last example, the dice in the first pool are sumed up and the ones in the second pool are considered distinctively, but there is a comma for both. Maybe keeping the brackets and replacing the comma with a + for the first pool would make things more clear. At least for me. Same for 1 and 2.

Also, if adding a bonus to each dice is really needed, it could maybe be written with parenthesis, but I would consider this as a different story/issue:
- 3(d20+5)>21: [25*, 12, 24*] = 2 successes (ie 20+5>21, 7+5<21 and 19+5>21)

Thanks for getting back on this and for the last release.

**Comment by GreenImp** [comment]
Thanks for the feedback, it's much appreciated.

I think with example 3, the second sub-roll should also be considered as a whole, so more like this:

```
{3d20+5,2d100}>21: {[20, 7, 19]+5*, [5, 22]*} = 2 successes (ie 20+7+19+5>21, 5+22>21)
```
Or another example:
```
{3d20+5,2d100}>21: {[20, 7, 19]+5*, [5, 19]*} = 2 successes (ie 20+7+19+5>21, 5+19>21)
```

Which makes it consistent with the others.

However, and this applies to the Keep and Drop modifiers as well, I'm not certain how we should show it visually.

If we use your example notation: `{3d20+5,2d100}>21`
A result like this seems wrong, as it implies that the `+5` has reached the target:
```
{[20, 7, 19]+5*, [5, 19]*}
```

One option I thought of was to surround sub-roll results in parenthesis, with the modifier "flag" on the outside:
```
{([20, 7, 19]+5)*, ([5, 19])*}
```
This could also be used for keep/drop modifiers:
```
{3d20+5, 2d100, 4d10}k2: {[20, 7, 19]+5, ([5, 19])d, [5, 7, 3, 10]}
```

Not sure if that's the best way of showing it - I'm open to suggestions.

**Comment by sigfriedmcwild** [comment]
I agree that we should keep it consistent between single sub roll and multiple sub rolls (and the single sub roll behaviour roll20 has makes no sense to me).

I like your proposed behaviours.

**Comment by vinceRV** [comment]
Yes, you are right about example 3. And parenthesis make things more easy to read.

👍

**Comment by GreenImp** [comment]
Thanks for the feedback guys. I'm making some progress on this, just laying the framework to allow me to get to groups working. Once that's done, I can start on the modifiers.

**Comment by GreenImp** [comment]
Just an progress update;
I've got group rolls working without modifiers. I'll start work on modifiers next.

At present, roll groups with no modifiers and single sub-rolls work much the same as non-grouped rolls:
```
{4d8}: {[3, 2, 7, 5]} = 17

// is equivalent to:
4d8: [3, 2, 7, 5] = 17
```

Multiple sub-rolls will get added together:
```
{4d8, 2d6}: {[3, 2, 7, 5], [1, 4]} = 22

// is equivalent to:
4d8+2d6: [3, 2, 7, 5]+[1, 4] = 22
```

And they work when used with other expressions as well:
```
(2d10/{4d8, 2d6})*2: ([9, 7]/{[3, 2, 7, 5], [1, 4]})*2 = 1.45
```

**Comment by GreenImp** [comment]
This is proving to be a lot more complicated than I thought, to get the modifiers working. I'm also not keen on the changes I've made to try and get it to work, as it feels a bit hacky.

I'm going to have to have a rethink about how to approach this one.

**Comment by GreenImp** [comment]
It's been a long time, but I've got roll groups working now! They're ready to go out in the next major release v5.0.0.

**Comment by SebaSOFT** [comment]
Awesome man, you did it!
Thanks for taking this request so far into the dev cycle... regards

**Comment by vinceRV** [comment]
Excellent! Looking forward to get it.
Thanks!

**Comment by GreenImp** [comment]
I've realised that there shouldn't be any breaking changes in this, so it can go out before I've finished other things. I've juts released this now in v4.5.0.

The documentation has also been updated: https://greenimp.github.io/rpg-dice-roller/guide/notation/group-rolls.html
